netbsd/Platform.c

```diff
@@ -323,7 +330,100 @@ bool Platform_getNetworkIO(NetworkIOData* data) {
 }
 
 void Platform_getBattery(double* percent, ACPresence* isOnAC) {
-   // TODO
-   (void)percent;
-   (void)isOnAC;
+   int fd;
+   bool isACAD, isBattery;
+   int64_t isPresent, isConnected;
+   int64_t curCharge, maxCharge;
+   int64_t totalCharge, totalCapacity;
+   prop_dictionary_t dict, fields, props;
+   prop_object_iterator_t devIter, fieldsIter;
+   prop_object_t device, class, fieldsArray, curValue, maxValue, descField;
+
+   totalCharge = 0;
+   totalCapacity = 0;
+
+   *percent = NAN;
+   *isOnAC = AC_ERROR;
+
+   fd = open(_PATH_SYSMON, O_RDONLY);
+   if (fd == -1)
+      goto error;
+
+   if (prop_dictionary_recv_ioctl(fd, ENVSYS_GETDICTIONARY, &dict) != 0)
+      goto error;
+
+   devIter = prop_dictionary_iterator(dict);
+   if (devIter == NULL)
+      goto error;
+
+   while ((device = prop_object_iterator_next(devIter)) != NULL) {
+      fieldsArray = prop_dictionary_get_keysym(dict, device);
+      if (fieldsArray == NULL)
+         goto error;
+
+      fieldsIter = prop_array_iterator(fieldsArray);
+      if (fieldsIter == NULL)
+         goto error;
+
+      isACAD = false;
+      isBattery = false;
+
+      /* only assume battery is not present if explicitly stated */
+      isPresent = 1;
+      isConnected = 0;
+      curCharge = 0;
+      maxCharge = 0;
+
+      while ((fields = prop_object_iterator_next(fieldsIter)) != NULL) {
+         props = prop_dictionary_get(fields, "device-properties");
+         if (props != NULL) {
+            class = prop_dictionary_get(props, "device-class");
+
+            /*
+             * After NetBSD 11's release NetBSD 9 will no longer be supported
+             * and these should be converted to prop_string_equals_string.
+             */
+
+            if (prop_string_equals_cstring(class, "ac-adapter")) {
+               isACAD = true;
+            } else if (prop_string_equals_cstring(class, "battery")) {
+               isBattery = true;
+            }
```

Comment:
Would checking for `prop_string_equals_string` work (using `configure.ac`)?

Thus having:
```c
#ifdef HAVE_PROP_STRING_EQUALS_STRING
#define proc_string_equals proc_string_equals_string
#else
#define proc_string_equals proc_string_equals_cstring
#endif
```

Or other question: How long is NetBSD 9 still to be supported, such that we could just ignore this old variant?

Reply:
Well, it's the current most recent release branch. proc_string_equals_cstring is only deprecated in the development branch.

Reply:
Does the non-deprecated version exist at all in the old version?

Reply:
It doesn't exist in 9.

Reply:
So we have to basically go via the check in `configure.ac` to make it compatible with both variants, without having to clean up the code a few years in …